src/sentry/hybridcloud/models/outbox.py

```diff
@@ -219,11 +220,143 @@ def process_shard(self, latest_shard_row: OutboxBase | None) -> Generator[Outbox
+        and bottom of the coalesce group.
+        2. We advance the schedule_for of all messages in the coalesce group. This
+        combined with the locks from 1. prevent concurrent deliveries of the same messages.
+        3. We commit the update transaction, and call the signal handler. Signal handler
+        need to be called outside of transactions as they frequently use RPC which
+        can have long latency. If a transaction was held during this time it could
+        hit statement timeouts and the transaction would be aborted.
+        4. Once the signal handler completes, we delete the coalesced messages
+        as they have been 'delivered'.
+
+        If no messages could be reserved or there was contention, this will return None.
+        """
+        using = router.db_for_write(type(self))
+        tags: dict[str, int | str] = {"category": "None", "synchronous": int(is_synchronous_flush)}
+
+        # If we have a latest_row, only operate on messages that were created before it.
+        filters = {} if latest_shard_row is None else dict(id__lte=latest_shard_row.id)
+        try:
+            now = timezone.now()
+            with transaction.atomic(using=using, savepoint=False):
+                # Lock all the rows we're going to process and attempt to reserve them.
+                reserved_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(**filters)
+                    .select_for_update(nowait=True)
+                    .order_by("id")
+                    .values_list("id", flat=True)
+                )
+                if not reserved_ids:
+                    yield None
+                    return
+
+                coalesced: OutboxBase = self.objects.get(id=reserved_ids[-1])
+                first_coalesced: OutboxBase = (
+                    self.objects.filter(id=reserved_ids[0]).first() or coalesced
+                )
+
+                # Reserve all messages in the coalesce group for a short time so they cannot
+                # be selected by concurrent calls.
+                self.objects.filter(
+                    id__in=reserved_ids,
+                ).update(scheduled_for=now + datetime.timedelta(minutes=5))
+        except OperationalError as e:
+            # If concurrent locking is happening on the table, gracefully pass and allow
+            # that work to process.
+            if "LockNotAvailable" in str(e):
+                yield None
+                return
+            else:
+                raise
+
+        tags["category"] = OutboxCategory(self.category).name
+        assert first_coalesced, "first_coalesced incorrectly set for non-empty coalesce group"
+        metrics.timing(
+            "outbox.coalesced_net_queue_time",
+            now.timestamp() - first_coalesced.date_added.timestamp(),
+            tags=tags,
+        )
+
+        # Yield the latest message for processing outside of a transaction
+        yield coalesced
```

Comment:
Do we check anywhere to ensure the coalesced group isn't already future scheduled?

So say we select and start processing a group of outboxes messages with IDs: `[1, 2]`, then enqueue another message with ID `3` and immediately attempt to process, won't this still allow us to process this outbox concurrently with `[1, 2]`?

Reply:
Added the `schedule_for` condition for when we're draining a single outbox shard.